File: internal/handlers/auth_handler.go
package handlers

import (
	"context"
	"crypto/rand"
	"fmt"
	"math/big"
	"net/http"
	"time"

	"chat-backend/internal/database"
	"chat-backend/internal/models"
	"chat-backend/internal/utils"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"golang.org/x/crypto/bcrypt"
)

type AuthHandler struct {
	db      *database.Database
	twilio  *utils.TwilioService
}

func NewAuthHandler(db *database.Database, twilio *utils.TwilioService) *AuthHandler {
	return &AuthHandler{
		db:     db,
		twilio: twilio,
	}
}

type RegisterRequest struct {
	PhoneNumber string `json:"phone_number" binding:"required"`
	Username    string `json:"username"`
	Password    string `json:"password"`
	UserType    string `json:"user_type"` // "normal" or "company"
	CompanyName string `json:"company_name,omitempty"`
	CompanyCategory string `json:"company_category,omitempty"`
}

type LoginRequest struct {
	PhoneNumber string `json:"phone_number" binding:"required"`
	Password    string `json:"password"`
}

type SendCodeRequest struct {
	PhoneNumber string `json:"phone_number" binding:"required"`
}

type VerifyCodeRequest struct {
	PhoneNumber string `json:"phone_number" binding:"required"`
	Code        string `json:"code" binding:"required"`
}

type RegisterWithCodeRequest struct {
	PhoneNumber     string `json:"phone_number" binding:"required"`
	Code            string `json:"code" binding:"required"`
	Username        string `json:"username"`
	UserType        string `json:"user_type"` // "normal" or "company"
	CompanyName     string `json:"company_name,omitempty"`
	CompanyCategory string `json:"company_category,omitempty"`
}

// verificationCodeDoc stores phone verification codes.
type verificationCodeDoc struct {
	PhoneNumber string    `bson:"phone_number"`
	Code        string    `bson:"code"`
	ExpiresAt   time.Time `bson:"expires_at"`
	CreatedAt   time.Time `bson:"created_at"`
}

// qrCodeCacheDoc stores QR code to user ID mapping for quick lookup.
type qrCodeCacheDoc struct {
	QRData   string    `bson:"qr_data"`
	UserID   string    `bson:"user_id"`
	CreatedAt time.Time `bson:"created_at"`
}

func (h *AuthHandler) storeVerificationCode(ctx context.Context, phone, code string, ttl time.Duration) error {
	expiresAt := time.Now().Add(ttl)
	// Store verification code in MongoDB
	doc := verificationCodeDoc{
		PhoneNumber: phone,
		Code:        code,
		ExpiresAt:   expiresAt,
		CreatedAt:   time.Now(),
	}
	_, err := h.db.MongoDB.Collection("verification_codes").InsertOne(ctx, doc)
	return err
}

func (h *AuthHandler) consumeVerificationCode(ctx context.Context, phone, code string) (bool, error) {
	// Check MongoDB: must match + not expired
	now := time.Now()
	filter := bson.M{
		"phone_number": phone,
		"code":         code,
		"expires_at":   bson.M{"$gt": now},
	}
	err := h.db.MongoDB.Collection("verification_codes").FindOne(ctx, filter).Err()
	if err == mongo.ErrNoDocuments {
		return false, nil
	}
	if err != nil {
		// Log MongoDB error for debugging
		fmt.Printf("Error finding verification code in MongoDB: %v\n", err)
		return false, err
	}
	// Consume: delete all codes for that phone (prevent reuse)
	_, _ = h.db.MongoDB.Collection("verification_codes").DeleteMany(ctx, bson.M{"phone_number": phone})
	return true, nil
}

func (h *AuthHandler) Register(c *gin.Context) {
	var req RegisterRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Check if user exists
	var existingUser models.User
	err := h.db.MongoDB.Collection("users").FindOne(
		context.Background(),
		bson.M{"phone_number": req.PhoneNumber},
	).Decode(&existingUser)

	if err == nil {
		c.JSON(http.StatusConflict, gin.H{"error": "User already exists"})
		return
	}

	// Hash password if provided (for future use)
	// Note: Password field is not yet in User model
	if req.Password != "" {
		_, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to hash password"})
			return
		}
		// TODO: Add password field to User model and store hashedPassword
	}

	// Generate QR code
	userID := primitive.NewObjectID()
	qrData, qrBase64, err := utils.GenerateQRCode(userID.Hex())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate QR code"})
		return
	}

	// Create user
	userType := req.UserType
	if userType == "" {
		userType = "normal"
	}
	user := models.User{
		ID:              userID,
		PhoneNumber:     req.PhoneNumber,
		QRCode:          qrBase64,
		Username:        req.Username,
		UserType:        userType,
		CompanyName:     req.CompanyName,
		CompanyCategory: req.CompanyCategory,
		IsAnonymous:     false,
		AccountStatus:   "active",
		CreatedAt:       time.Now(),
		UpdatedAt:       time.Now(),
		LastActive:      time.Now(),
	}

	_, err = h.db.MongoDB.Collection("users").InsertOne(context.Background(), user)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create user"})
		return
	}

	// Store QR data in MongoDB for quick lookup
	qrCacheDoc := qrCodeCacheDoc{
		QRData:    qrData,
		UserID:    userID.Hex(),
		CreatedAt: time.Now(),
	}
	// Use upsert to update if exists
	filter := bson.M{"qr_data": qrData}
	update := bson.M{"$set": qrCacheDoc}
	_, _ = h.db.MongoDB.Collection("qr_code_cache").UpdateOne(context.Background(), filter, update, options.Update().SetUpsert(true))

	// Generate token
	token, err := utils.GenerateToken(userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate token"})
		return
	}

	c.JSON(http.StatusCreated, gin.H{
		"token": token,
		"user":  user,
		"qr":    qrBase64,
	})
}

func (h *AuthHandler) Login(c *gin.Context) {
	var req LoginRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var user models.User
	err := h.db.MongoDB.Collection("users").FindOne(
		context.Background(),
		bson.M{"phone_number": req.PhoneNumber},
	).Decode(&user)

	if err == mongo.ErrNoDocuments {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid credentials"})
		return
	}

	// In a real app, verify password here
	// For now, we'll just check if user exists

	token, err := utils.GenerateToken(user.ID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate token"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"token": token,
		"user":  user,
	})
}

func (h *AuthHandler) GetQRCode(c *gin.Context) {
	userIDStr := c.Param("user_id")
	userID, err := primitive.ObjectIDFromHex(userIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid user ID"})
		return
	}

	var user models.User
	err = h.db.MongoDB.Collection("users").FindOne(
		context.Background(),
		bson.M{"_id": userID},
	).Decode(&user)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"qr_code": user.QRCode})
}

// SendCode sends a verification code to the phone number
func (h *AuthHandler) SendCode(c *gin.Context) {
	var req SendCodeRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request: " + err.Error()})
		return
	}

	// Validate phone number
	if req.PhoneNumber == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Phone number is required"})
		return
	}

	// Basic phone number validation (should start with + and have at least 10 digits)
	if len(req.PhoneNumber) < 10 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid phone number format"})
		return
	}

	// Generate 6-digit code
	n, err := rand.Int(rand.Reader, big.NewInt(1000000))
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate code"})
		return
	}
	code := fmt.Sprintf("%06d", n.Int64())

	// Store code with 5 minute expiration (Redis preferred, Mongo fallback)
	if err := h.storeVerificationCode(context.Background(), req.PhoneNumber, code, 5*time.Minute); err != nil {
		// If storage fails, we cannot safely verify later
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to store verification code"})
		return
	}

	// Send SMS via Twilio
	twilioSent := false
	if h.twilio != nil && h.twilio.IsEnabled() {
		err = h.twilio.SendVerificationCode(req.PhoneNumber, code)
		if err != nil {
			// Log error but don't fail the request - code is still stored in Redis
			fmt.Printf("Failed to send SMS via Twilio: %v\n", err)
		} else {
			twilioSent = true
		}
	}

	// Response
	response := gin.H{
		"message": "Verification code sent",
		"success": true,
	}

	// Only return code in development mode (when Twilio is disabled or failed)
	// In production with Twilio enabled, don't return the code
	if h.twilio == nil || !h.twilio.IsEnabled() || !twilioSent {
		response["code"] = code // Development mode or Twilio failed
	}

	c.JSON(http.StatusOK, response)
}

// VerifyCode verifies the code for login
func (h *AuthHandler) VerifyCode(c *gin.Context) {
	var req VerifyCodeRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Create context with timeout for database operations
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	ok, err := h.consumeVerificationCode(ctx, req.PhoneNumber, req.Code)
	if err != nil {
		fmt.Printf("Error consuming verification code: %v\n", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Verification lookup failed"})
		return
	}
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid or expired code"})
		return
	}

	// Find user
	var user models.User
	err = h.db.MongoDB.Collection("users").FindOne(
		ctx,
		bson.M{"phone_number": req.PhoneNumber},
	).Decode(&user)

	if err == mongo.ErrNoDocuments {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found. Please register first."})
		return
	}

	if err != nil {
		// Log the actual error for debugging
		fmt.Printf("Error finding user in MongoDB: %v\n", err)
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Database error",
			"details": err.Error(),
		})
		return
	}

	// Generate token
	token, err := utils.GenerateToken(user.ID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate token"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"token": token,
		"user":  user,
	})
}

// RegisterWithCode registers a new user after code verification
func (h *AuthHandler) RegisterWithCode(c *gin.Context) {
	var req RegisterWithCodeRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Validate user type
	if req.UserType != "normal" && req.UserType != "company" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "user_type must be 'normal' or 'company'"})
		return
	}

	// Validate company fields if user type is company
	if req.UserType == "company" {
		if req.CompanyName == "" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "company_name is required for company users"})
			return
		}
		if req.CompanyCategory == "" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "company_category is required for company users"})
			return
		}
	}

	// Verify code
	ok, err := h.consumeVerificationCode(context.Background(), req.PhoneNumber, req.Code)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Verification lookup failed"})
		return
	}
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid or expired code"})
		return
	}

	// Check if user already exists
	var existingUser models.User
	err = h.db.MongoDB.Collection("users").FindOne(
		context.Background(),
		bson.M{"phone_number": req.PhoneNumber},
	).Decode(&existingUser)

	if err == nil {
		c.JSON(http.StatusConflict, gin.H{"error": "User already exists"})
		return
	}

	// Generate QR code
	userID := primitive.NewObjectID()
	qrData, qrBase64, err := utils.GenerateQRCode(userID.Hex())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate QR code"})
		return
	}

	// Create user
	user := models.User{
		ID:              userID,
		PhoneNumber:     req.PhoneNumber,
		QRCode:          qrBase64,
		Username:        req.Username,
		UserType:        req.UserType,
		CompanyName:     req.CompanyName,
		CompanyCategory: req.CompanyCategory,
		IsAnonymous:     false,
		AccountStatus:   "active",
		CreatedAt:       time.Now(),
		UpdatedAt:       time.Now(),
		LastActive:      time.Now(),
	}

	_, err = h.db.MongoDB.Collection("users").InsertOne(context.Background(), user)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create user"})
		return
	}

	// Store QR data in MongoDB for quick lookup
	qrCacheDoc := qrCodeCacheDoc{
		QRData:    qrData,
		UserID:    userID.Hex(),
		CreatedAt: time.Now(),
	}
	// Use upsert to update if exists
	filter := bson.M{"qr_data": qrData}
	update := bson.M{"$set": qrCacheDoc}
	_, _ = h.db.MongoDB.Collection("qr_code_cache").UpdateOne(context.Background(), filter, update, options.Update().SetUpsert(true))

	// Generate token
	token, err := utils.GenerateToken(userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate token"})
		return
	}

	c.JSON(http.StatusCreated, gin.H{
		"token": token,
		"user":  user,
		"qr":    qrBase64,
	})
}

func (h *AuthHandler) VerifyPhone(c *gin.Context) {
	// Legacy endpoint - redirects to SendCode
	h.SendCode(c)
}

File: src/handlers/authHandler.js
'use strict';

const { ObjectId } = require('mongodb');
const bcrypt = require('bcryptjs');
const crypto = require('crypto');
const { getDB } = require('../database');
const { generateToken } = require('../utils/jwt');
const { generateQRCode } = require('../utils/qrcode');

// ─── Helpers ──────────────────────────────────────────────────────────────────

async function storeVerificationCode(phone, code, ttlMs) {
  const db = getDB();
  const expiresAt = new Date(Date.now() + ttlMs);
  await db.collection('verification_codes').insertOne({
    phone_number: phone,
    code,
    expires_at: expiresAt,
    created_at: new Date(),
  });
}

async function checkVerificationCode(phone, code) {
  const db = getDB();
  const doc = await db.collection('verification_codes').findOne({
    phone_number: phone,
    code,
    expires_at: { $gt: new Date() },
  });
  return !!doc;
}

async function consumeVerificationCode(phone, code) {
  const valid = await checkVerificationCode(phone, code);
  if (!valid) return false;
  const db = getDB();
  await db.collection('verification_codes').deleteMany({ phone_number: phone });
  return true;
}

// ─── Handlers ─────────────────────────────────────────────────────────────────

async function register(req, res) {
  try {
    const { phone_number, username, password, user_type, company_name, company_category } = req.body;
    if (!phone_number) return res.status(400).json({ error: 'phone_number is required' });

    const db = getDB();

    const existing = await db.collection('users').findOne({ phone_number });
    if (existing) return res.status(409).json({ error: 'User already exists' });

    // Hash password if provided (for future use)
    if (password) {
      await bcrypt.hash(password, 10); // stored if password field added to model
    }

    const userID = new ObjectId();
    const { qrData, qrBase64 } = await generateQRCode(userID.toHexString());

    const userType = user_type || 'normal';
    const now = new Date();
    const user = {
      _id: userID,
      phone_number,
      qr_code: qrBase64,
      username: username || '',
      user_type: userType,
      company_name: company_name || '',
      company_category: company_category || '',
      is_anonymous: false,
      account_status: 'active',
      created_at: now,
      updated_at: now,
      last_active: now,
    };

    await db.collection('users').insertOne(user);

    // Cache QR data
    await db.collection('qr_code_cache').updateOne(
      { qr_data: qrData },
      { $set: { qr_data: qrData, user_id: userID.toHexString(), created_at: now } },
      { upsert: true }
    );

    const token = generateToken(userID.toHexString());

    return res.status(201).json({ token, user, qr: qrBase64 });
  } catch (err) {
    console.error('register error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function login(req, res) {
  try {
    const { phone_number } = req.body;
    if (!phone_number) return res.status(400).json({ error: 'phone_number is required' });

    const db = getDB();
    const user = await db.collection('users').findOne({ phone_number });
    if (!user) return res.status(401).json({ error: 'Invalid credentials' });

    const token = generateToken(user._id.toHexString());
    return res.status(200).json({ token, user });
  } catch (err) {
    console.error('login error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getQRCode(req, res) {
  try {
    const { user_id } = req.params;
    let userIDObj;
    try { userIDObj = new ObjectId(user_id); } catch { return res.status(400).json({ error: 'Invalid user ID' }); }

    const db = getDB();
    const user = await db.collection('users').findOne({ _id: userIDObj });
    if (!user) return res.status(404).json({ error: 'User not found' });

    return res.status(200).json({ qr_code: user.qr_code });
  } catch (err) {
    console.error('getQRCode error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function sendCode(req, res) {
  try {
    const { phone_number } = req.body;
    if (!phone_number || phone_number.length < 10) {
      return res.status(400).json({ error: 'Invalid phone number format' });
    }

    // Generate 6-digit code
    const code = String(crypto.randomInt(0, 1000000)).padStart(6, '0');

    // Store with 5 minute TTL
    await storeVerificationCode(phone_number, code, 5 * 60 * 1000);

    // Send SMS via Twilio
    let twilioSent = false;
    const twilioService = req.app.get('twilioService');
    if (twilioService && twilioService.isEnabled()) {
      try {
        await twilioService.sendVerificationCode(phone_number, code);
        twilioSent = true;
      } catch (twilioErr) {
        console.error('Failed to send SMS via Twilio:', twilioErr.message);
      }
    }

    const response = { message: 'Verification code sent', success: true };
    if (!twilioService || !twilioService.isEnabled() || !twilioSent) {
      response.code = code; // development mode
    }

    return res.status(200).json(response);
  } catch (err) {
    console.error('sendCode error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function verifyCode(req, res) {
  try {
    const { phone_number, code } = req.body;
    if (!phone_number || !code) {
      return res.status(400).json({ error: 'phone_number and code are required' });
    }

    // Check code without consuming
    const ok = await checkVerificationCode(phone_number, code);
    if (!ok) return res.status(401).json({ error: 'Invalid or expired code' });

    const db = getDB();
    const user = await db.collection('users').findOne({ phone_number });

    if (!user) {
      // User not found — do NOT consume code (it will be used for register-with-code)
      return res.status(404).json({ error: 'User not found. Please register first.' });
    }

    // Consume code on successful login
    await consumeVerificationCode(phone_number, code);

    const token = generateToken(user._id.toHexString());
    return res.status(200).json({ token, user });
  } catch (err) {
    console.error('verifyCode error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function registerWithCode(req, res) {
  try {
    const {
      phone_number, code, username, profession,
      user_type, company_name, company_category,
    } = req.body;

    if (!phone_number || !code) {
      return res.status(400).json({ error: 'phone_number and code are required' });
    }

    if (user_type !== 'normal' && user_type !== 'company') {
      return res.status(400).json({ error: "user_type must be 'normal' or 'company'" });
    }

    if (user_type === 'company') {
      if (!company_name) return res.status(400).json({ error: 'company_name is required for company users' });
      if (!company_category) return res.status(400).json({ error: 'company_category is required for company users' });
    }

    // Verify and consume code
    const ok = await consumeVerificationCode(phone_number, code);
    if (!ok) return res.status(401).json({ error: 'Invalid or expired code' });

    const db = getDB();
    const existing = await db.collection('users').findOne({ phone_number });
    if (existing) return res.status(409).json({ error: 'User already exists' });

    const userID = new ObjectId();
    const { qrData, qrBase64 } = await generateQRCode(userID.toHexString());

    const now = new Date();
    const user = {
      _id: userID,
      phone_number,
      qr_code: qrBase64,
      username: username || '',
      profession: profession || '',
      user_type: user_type || 'normal',
      company_name: company_name || '',
      company_category: company_category || '',
      is_anonymous: false,
      account_status: 'active',
      created_at: now,
      updated_at: now,
      last_active: now,
    };

    await db.collection('users').insertOne(user);

    await db.collection('qr_code_cache').updateOne(
      { qr_data: qrData },
      { $set: { qr_data: qrData, user_id: userID.toHexString(), created_at: now } },
      { upsert: true }
    );

    const token = generateToken(userID.toHexString());
    return res.status(201).json({ token, user, qr: qrBase64 });
  } catch (err) {
    console.error('registerWithCode error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

// Legacy endpoint — same as sendCode
const verifyPhone = sendCode;

module.exports = {
  register,
  login,
  getQRCode,
  sendCode,
  verifyCode,
  registerWithCode,
  verifyPhone,
};
